#### Multinomial Naive Bayes

It is a kind of naive bayes algorithm that works with the text.

#### Theory:--

1.  It helps in building the vocabulary for each and every word.

2.  It then performs the matrix o all the word counts.

3.  Then it finds the frequencies of every word in the text.

##### Main aim of the algorithm:---

(a.)    The main aim of the algorithm is to predict the type of the Iris Plant Species using Gaussian Naive Bayes

(b.)    We can even improve the accuarcy of the model using Hyper Parameter Tuning with GridSearchCV



##### Step 1:  Import  all  the necessary libraries

In [1]:
import  numpy               as  np
import  pandas              as  pd
import  matplotlib.pyplot   as  plt
import  seaborn             as  sns

from    sklearn.model_selection         import       train_test_split,  GridSearchCV,   StratifiedKFold
from    sklearn.preprocessing           import       StandardScaler
from    sklearn.feature_extraction.text import       CountVectorizer
from    sklearn.naive_bayes             import       MultinomialNB
from    sklearn.metrics                 import       accuracy_score, confusion_matrix, classification_report

#### Step 2:    Load the dataset

In [2]:
df = pd.read_csv("email_data.csv",encoding='latin-1')

In [3]:
df.head()

,v1,v2,Unnamed: 2,Unnamed: 3,Unnamed: 4
0,ham,"Go until jurong point, crazy.. Available only ...",NaN,NaN,NaN
1,ham,Ok lar... Joking wif u oni...,NaN,NaN,NaN
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...,NaN,NaN,NaN
3,ham,U dun say so early hor... U c already then say...,NaN,NaN,NaN
4,ham,"Nah I don't think he goes to usf, he lives aro...",NaN,NaN,NaN


#### OBSERVATIONS:

1.  The above dataset contains the text and the type of label associated with the text whether it is "ham" or "spam".

#### Step 3: Perform the Data Exploration

In [4]:
#### get the top five records of the dataset

df.head()

,v1,v2,Unnamed: 2,Unnamed: 3,Unnamed: 4
0,ham,"Go until jurong point, crazy.. Available only ...",NaN,NaN,NaN
1,ham,Ok lar... Joking wif u oni...,NaN,NaN,NaN
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...,NaN,NaN,NaN
3,ham,U dun say so early hor... U c already then say...,NaN,NaN,NaN
4,ham,"Nah I don't think he goes to usf, he lives aro...",NaN,NaN,NaN


In [5]:
#### get the bottom five records of the dataset

df.tail()

,v1,v2,Unnamed: 2,Unnamed: 3,Unnamed: 4
5567,spam,This is the 2nd time we have tried 2 contact u...,NaN,NaN,NaN
5568,ham,Will Ì_ b going to esplanade fr home?,NaN,NaN,NaN
5569,ham,"Pity, * was in mood for that. So...any other s...",NaN,NaN,NaN
5570,ham,The guy did some bitching but I acted like i'd...,NaN,NaN,NaN
5571,ham,Rofl. Its true to its name,NaN,NaN,NaN


In [6]:
#### get the length of the dataset

print("Length of the dataset is:", len(df))

Length of the dataset is: 5572


In [7]:
#### get the shape of the dataset

print("Shape of the dataset is:", df.shape)

Shape of the dataset is: (5572, 5)


In [8]:
#### get the columns used in the dataset

df.columns

Index(['v1', 'v2', 'Unnamed: 2', 'Unnamed: 3', 'Unnamed: 4'], dtype='object')

In [9]:
#### get the information about the columns in the dataset

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5572 entries, 0 to 5571
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   v1          5572 non-null   object
 1   v2          5572 non-null   object
 2   Unnamed: 2  50 non-null     object
 3   Unnamed: 3  12 non-null     object
 4   Unnamed: 4  6 non-null      object
dtypes: object(5)
memory usage: 217.8+ KB


#### OBSERVATIONS:

1.  All the columns are string in nature.

In [10]:
#### get the descriptive summary statistics about each column in the dataset

df.describe()

,v1,v2,Unnamed: 2,Unnamed: 3,Unnamed: 4
count,5572,5572,50,12,6
unique,2,5169,43,10,5
top,ham,"Sorry, I'll call later","bt not his girlfrnd... G o o d n i g h t . . .@""","MK17 92H. 450Ppw 16""","GNT:-)"""
freq,4825,30,3,2,2


#### Step 4:    Perform the Data Cleaning and Feature Engineering

In [11]:
df.head()

,v1,v2,Unnamed: 2,Unnamed: 3,Unnamed: 4
0,ham,"Go until jurong point, crazy.. Available only ...",NaN,NaN,NaN
1,ham,Ok lar... Joking wif u oni...,NaN,NaN,NaN
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...,NaN,NaN,NaN
3,ham,U dun say so early hor... U c already then say...,NaN,NaN,NaN
4,ham,"Nah I don't think he goes to usf, he lives aro...",NaN,NaN,NaN


In [12]:
#### Remove all the NULL Columns from the dataset

df.drop(columns={"Unnamed: 2","Unnamed: 3","Unnamed: 4"},inplace=True)

In [13]:
df.head()

,v1,v2
0,ham,"Go until jurong point, crazy.. Available only ..."
1,ham,Ok lar... Joking wif u oni...
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...
3,ham,U dun say so early hor... U c already then say...
4,ham,"Nah I don't think he goes to usf, he lives aro..."


In [14]:
#### Rename the column names to their proper names

df.rename(columns={"v1":"label","v2":"text"},inplace=True)

In [15]:
df.head()

,label,text
0,ham,"Go until jurong point, crazy.. Available only ..."
1,ham,Ok lar... Joking wif u oni...
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...
3,ham,U dun say so early hor... U c already then say...
4,ham,"Nah I don't think he goes to usf, he lives aro..."


In [16]:
#### Assign the column names to their proper positions

df = df[["text", "label"]]

In [17]:
df.head()

,text,label
0,"Go until jurong point, crazy.. Available only ...",ham
1,Ok lar... Joking wif u oni...,ham
2,Free entry in 2 a wkly comp to win FA Cup fina...,spam
3,U dun say so early hor... U c already then say...,ham
4,"Nah I don't think he goes to usf, he lives aro...",ham


#### OBSERVATIONS:

1.  Now the column names have been assigned to their proper index positions.

In [18]:
#### To check for all NULL records in the dataset

df.isnull().sum()

text     0
label    0
dtype: int64

#### OBSERVATIONS:

1.  There are no NULL records in the dataset.

In [19]:
#### To check for the duplicate records in the dataset

df[df.duplicated()]

,text,label
102,As per your request 'Melle Melle (Oru Minnamin...,ham
153,As per your request 'Melle Melle (Oru Minnamin...,ham
206,"As I entered my cabin my PA said, '' Happy B'd...",ham
222,"Sorry, I'll call later",ham
325,No calls..messages..missed calls,ham
...,...,...
5524,You are awarded a SiPix Digital Camera! call 0...,spam
5535,"I know you are thinkin malaria. But relax, chi...",ham
5539,Just sleeping..and surfing,ham
5553,Hahaha..use your brain dear,ham


#### OBSERVATIONS:

1.There are total 403 duplicate records in the dataset, so remove it.

In [20]:
#### Remove the duplicate records from the dataset

df = df.drop_duplicates()

In [21]:
#### To again check for the duplicate records in the dataset

df[df.duplicated()]

,text,label


#### OBSERVATIONS:

1.  There are no duplicate records in the dataset.

In [22]:
df.head()

,text,label
0,"Go until jurong point, crazy.. Available only ...",ham
1,Ok lar... Joking wif u oni...,ham
2,Free entry in 2 a wkly comp to win FA Cup fina...,spam
3,U dun say so early hor... U c already then say...,ham
4,"Nah I don't think he goes to usf, he lives aro...",ham


In [23]:
#### Convert the label values into their numerical form

df['label'] = df['label'].map({"spam":1, "ham":0})

In [24]:
df.head()

,text,label
0,"Go until jurong point, crazy.. Available only ...",0
1,Ok lar... Joking wif u oni...,0
2,Free entry in 2 a wkly comp to win FA Cup fina...,1
3,U dun say so early hor... U c already then say...,0
4,"Nah I don't think he goes to usf, he lives aro...",0


#### OBSERVATIONS:

1.  Now all the values in the label column have been converted into their numerical values.

#### Step 5:   Divide the dataset into independent and dependent features

In [25]:
#### independent features

X = df['text']

print(X)

0       Go until jurong point, crazy.. Available only ...
1                           Ok lar... Joking wif u oni...
2       Free entry in 2 a wkly comp to win FA Cup fina...
3       U dun say so early hor... U c already then say...
4       Nah I don't think he goes to usf, he lives aro...
                              ...                        
5567    This is the 2nd time we have tried 2 contact u...
5568                Will Ì_ b going to esplanade fr home?
5569    Pity, * was in mood for that. So...any other s...
5570    The guy did some bitching but I acted like i'd...
5571                           Rofl. Its true to its name
Name: text, Length: 5169, dtype: object


In [26]:
#### dependent features

Y = df['label']

print(Y)

0       0
1       0
2       1
3       0
4       0
       ..
5567    1
5568    0
5569    0
5570    0
5571    0
Name: label, Length: 5169, dtype: int64


#### Step 6: Divide the independent and dependent features in the dataset into training and testing data

In [27]:
from  sklearn.model_selection  import  train_test_split

X_train, X_test, Y_train, Y_test = train_test_split(X,Y,test_size=0.2,random_state=42,stratify=Y)

In [28]:
X_train

336     Cool. So how come you havent been wined and di...
390                             Huh so late... Fr dinner?
583     my ex-wife was not able to have kids. Do you w...
1387    Hope ur head doesn't hurt 2 much ! Am ploughin...
1095                              Ryder unsold.now gibbs.
                              ...                        
606     XCLUSIVE@CLUBSAISAI 2MOROW 28/5 SOIREE SPECIAL...
105     Thanks a lot for your wishes on my birthday. T...
3789    I love you !!! You know? Can you feel it? Does...
1911    Lol grr my mom is taking forever with my presc...
3742    Nah, Wednesday. When should I bring the mini c...
Name: text, Length: 4135, dtype: object

In [29]:
X_test

5318    Good morning, my Love ... I go to sleep now an...
5281               And how you will do that, princess? :)
2459       Nope. Meanwhile she talk say make i greet you.
1305    Designation is software developer and may be s...
5083    Aiya we discuss later lar... Pick Ì_ up at 4 i...
                              ...                        
1119    Bored of speed dating? Try SPEEDCHAT, txt SPEE...
3068                   Ay wana meet on sat?Ì_ wkg on sat?
812     Congratulations ur awarded either å£500 of CD ...
1560    Just got some gas money, any chance you and th...
1238    Dear relieved of westonzoyland, all going to p...
Name: text, Length: 1034, dtype: object

In [30]:
print("Shape of the input training data is:", X_train.shape)
print("Shape of the input testing  data is:", X_test.shape)

Shape of the input training data is: (4135,)
Shape of the input testing  data is: (1034,)


In [31]:
Y_train

336     0
390     0
583     0
1387    0
1095    0
       ..
606     1
105     0
3789    0
1911    0
3742    0
Name: label, Length: 4135, dtype: int64

In [32]:
Y_test

5318    0
5281    0
2459    0
1305    0
5083    0
       ..
1119    1
3068    0
812     1
1560    0
1238    0
Name: label, Length: 1034, dtype: int64

In [33]:
print("Shape of the output training data is:", Y_train.shape)
print("Shape of the output testing  data is:", Y_test.shape)

Shape of the output training data is: (4135,)
Shape of the output testing  data is: (1034,)


#### Step 7: Convert the input text into numerical vectors

In [34]:
from  sklearn.feature_extraction.text  import  CountVectorizer


#### create an object for CountVectorizer

count = CountVectorizer(stop_words="english")


#### using the object for CountVectorizer, scale the inputs

X_train_scaled = count.fit_transform(X_train)

print(X_train_scaled)

<Compressed Sparse Row sparse matrix of dtype 'int64'
	with 31574 stored elements and shape (4135, 7329)>
  Coords	Values
  (0, 1928)	1
  (0, 1831)	1
  (0, 3212)	1
  (0, 7113)	1
  (0, 2239)	1
  (1, 3391)	1
  (1, 3850)	1
  (1, 2840)	1
  (1, 2243)	1
  (2, 2571)	1
  (2, 7096)	1
  (2, 749)	1
  (2, 3760)	2
  (2, 6984)	1
  (2, 2097)	1
  (3, 1831)	1
  (3, 3342)	1
  (3, 6821)	1
  (3, 3221)	1
  (3, 2297)	1
  (3, 3407)	1
  (3, 4990)	1
  (3, 7016)	1
  (3, 4952)	1
  (3, 3575)	1
  :	:
  (4132, 2295)	2
  (4132, 5755)	1
  (4132, 3776)	1
  (4132, 812)	1
  (4132, 6991)	1
  (4132, 2352)	1
  (4132, 1253)	1
  (4132, 856)	1
  (4133, 3932)	1
  (4133, 4005)	1
  (4133, 6350)	1
  (4133, 2813)	1
  (4133, 1131)	1
  (4133, 6744)	1
  (4133, 4307)	1
  (4133, 4367)	1
  (4133, 3120)	1
  (4133, 5113)	1
  (4133, 4917)	1
  (4134, 1444)	1
  (4134, 7037)	1
  (4134, 1162)	1
  (4134, 4470)	1
  (4134, 4296)	1
  (4134, 1689)	1


#### OBSERVATIONS:

1.  All the input text in the X_train has been converted into the matrix of the numerical word counts.

In [35]:
X_test_scaled = count.transform(X_test)

print(X_test_scaled)

<Compressed Sparse Row sparse matrix of dtype 'int64'
	with 6868 stored elements and shape (1034, 7329)>
  Coords	Values
  (0, 1149)	1
  (0, 1271)	1
  (0, 2097)	1
  (0, 2688)	1
  (0, 3048)	1
  (0, 3099)	1
  (0, 3776)	1
  (0, 4046)	2
  (0, 4389)	1
  (0, 4723)	1
  (0, 5899)	1
  (0, 6506)	1
  (0, 7125)	1
  (1, 5135)	1
  (2, 3104)	1
  (2, 4137)	1
  (2, 4591)	1
  (2, 5628)	1
  (2, 6353)	1
  (3, 1691)	1
  (3, 5968)	1
  (4, 867)	1
  (4, 2265)	1
  (4, 3843)	1
  (4, 3853)	1
  :	:
  (1031, 2348)	1
  (1031, 2508)	1
  (1031, 2853)	1
  (1031, 2996)	1
  (1031, 3870)	1
  (1031, 4451)	1
  (1031, 6570)	1
  (1031, 6722)	1
  (1031, 6821)	1
  (1031, 6934)	1
  (1031, 7044)	1
  (1031, 7217)	1
  (1032, 1647)	1
  (1032, 2938)	1
  (1032, 2946)	1
  (1032, 3064)	1
  (1032, 3085)	1
  (1032, 3707)	1
  (1032, 4374)	1
  (1032, 4495)	1
  (1032, 6984)	1
  (1033, 2107)	1
  (1033, 2479)	1
  (1033, 3035)	1
  (1033, 4969)	1


#### OBSERVATIONS:

1.  The Count Vectorizer has been applied on the input text.

2.  Then a vocabulary of all the word in the input text is build.

3.  Each and every word is then converted into the matrix of the word counts.

4.  The above output shows the matrix representation of each and every word along with its frequency.

5.  The input text in X_test has been converted into the matrix of the word counts.

#### Step 8: Train the Multinomial Naive Bayes

In [36]:
from  sklearn.naive_bayes import MultinomialNB

#### Create an object for Multinomial Naive Bayes

mnb = MultinomialNB()

#### using the object for Multinomial Naive Bayes

mnb.fit(X_train_scaled, Y_train)

MultinomialNB()

#### Step 9: Predict the Model

In [37]:
Y_pred = mnb.predict(X_test_scaled)

In [38]:
Y_pred

array([0, 0, 0, ..., 1, 0, 0])

#### Step 10: Evaluate the Classification Metrics

In [39]:
from        sklearn.metrics             import  accuracy_score, confusion_matrix, classification_report

print("===============================Accuracy score=======================================================")

ac = accuracy_score(Y_test, Y_pred)*100.0

print("Accuracy score of the model is:", ac)



print("===============================Confusion matrix=======================================================")

cm = confusion_matrix(Y_test, Y_pred)

print("Confusion matrix of the model is:", cm)



print("===============================Classification report=======================================================")

cr = classification_report(Y_test, Y_pred)

print("Classification report of the model is:", cr)

===============================Accuracy score=======================================================
Accuracy score of the model is: 98.45261121856866
===============================Confusion matrix=======================================================
Confusion matrix of the model is: [[898   5]
 [ 11 120]]
===============================Classification report=======================================================
Classification report of the model is:               precision    recall  f1-score   support

           0       0.99      0.99      0.99       903
           1       0.96      0.92      0.94       131

    accuracy                           0.98      1034
   macro avg       0.97      0.96      0.96      1034
weighted avg       0.98      0.98      0.98      1034



#### OBSERVATIONS:

1.  The accuracy of email spam classiication on the email dataset using Multinomial Naive Bayes is 98.45. It is very high.

2.  The model is working fine.

#### Step 11: Predict whether the email is spam or ham based on the sample data

In [40]:
#### define the sample messages

sample_messages = [
    "Congratulations! You've won a free iPhone. Click here to claim.",
    "Hey, are we still meeting for lunch today?"
]

print(sample_messages)

#### perform the vectorization on the input text

scaled_inputs = count.transform(sample_messages)

print(scaled_inputs)


#### perform the prediction on the vectorized data

prediction = mnb.predict(scaled_inputs)

print(prediction)


print("==========================First Sentence===============================================")

if(prediction[0] == 1):
    print("The email message is Spam")
else:
    print("The email message is Ham")


print("==========================Second Sentence===============================================")

if(prediction[1] == 1):
    print("The email message is Spam")
else:
    print("The email message is Ham")

["Congratulations! You've won a free iPhone. Click here to claim.", 'Hey, are we still meeting for lunch today?']
<Compressed Sparse Row sparse matrix of dtype 'int64'
	with 11 stored elements and shape (2, 7329)>
  Coords	Values
  (0, 1748)	1
  (0, 1766)	1
  (0, 1888)	1
  (0, 2853)	1
  (0, 3569)	1
  (0, 6879)	1
  (0, 7157)	1
  (1, 3273)	1
  (1, 4075)	1
  (1, 4236)	1
  (1, 6576)	1
[1 0]
==========================First Sentence===============================================
The email message is Spam
==========================Second Sentence===============================================
The email message is Ham


#### Step 12: Perform Hyper Parameter Tuning using GridSearchCV

In [41]:
#### define the pipeline

from sklearn.pipeline  import  Pipeline

#### create the object for pipeline

pipeline = Pipeline([
    ("count"    ,   CountVectorizer())  ,
    ("mnb"      ,   MultinomialNB())
])

In [42]:
#### define the param grid

param_grid = {
    "count__ngram_range"        :   [(1,1),(1,2)]              ,           #### unigram + bigram
    "count__max_df"             :   [0.75, 0.9, 1.0]           ,           #### remove most frequent words
    "count__min_df"             :   [1,2,5]                    ,           #### remove the rare words
    "count__stop_words"         :   [None,"english"]           ,           #### remove the words that have no  meaning
    "mnb__alpha"                :   [0.1, 0.5, 1.0, 2.0]                   #### smoothing parameter 
}

In [43]:
#### define the cross validation

cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

In [44]:
#### create the object for GridSearchCV

grid = GridSearchCV(
    pipeline                            ,
    param_grid      =       param_grid  ,
    cv              =          cv       ,
    scoring         =       'accuracy'  ,
    n_jobs          =          -1       ,
    verbose         =           2 
)

In [45]:
#### train the grid search cv model

grid.fit(X_train, Y_train)

Fitting 5 folds for each of 144 candidates, totalling 720 fits


GridSearchCV(cv=StratifiedKFold(n_splits=5, random_state=42, shuffle=True),
             estimator=Pipeline(steps=[('count', CountVectorizer()),
                                       ('mnb', MultinomialNB())]),
             n_jobs=-1,
             param_grid={'count__max_df': [0.75, 0.9, 1.0],
                         'count__min_df': [1, 2, 5],
                         'count__ngram_range': [(1, 1), (1, 2)],
                         'count__stop_words': [None, 'english'],
                         'mnb__alpha': [0.1, 0.5, 1.0, 2.0]},
             scoring='accuracy', verbose=2)

In [46]:
#### Obtain the best parameters and the estimator of the model

best_parameters = grid.best_params_

print("Best Parameter of the model is:", best_parameters)


best_estimator = grid.best_estimator_

print("Best Estimator of the model is:", best_estimator)

Best Parameter of the model is: {'count__max_df': 0.75, 'count__min_df': 1, 'count__ngram_range': (1, 2), 'count__stop_words': 'english', 'mnb__alpha': 1.0}
Best Estimator of the model is: Pipeline(steps=[('count',
                 CountVectorizer(max_df=0.75, ngram_range=(1, 2),
                                 stop_words='english')),
                ('mnb', MultinomialNB())])


In [47]:
#### predict the model

Y_pred_grid = best_estimator.predict(X_test)

In [48]:
Y_pred_grid

array([0, 0, 0, ..., 1, 0, 0])

In [49]:
#### Evaluate the Classification Metrics


from        sklearn.metrics             import  accuracy_score, confusion_matrix, classification_report

print("===============================Accuracy score=======================================================")

acgrid = accuracy_score(Y_test, Y_pred_grid)*100.0

print("Accuracy score of the model is:", acgrid)



print("===============================Confusion matrix=======================================================")

cmgrid = confusion_matrix(Y_test, Y_pred_grid)

print("Confusion matrix of the model is:", cmgrid)



print("===============================Classification report=======================================================")

crgrid = classification_report(Y_test, Y_pred_grid)

print("Classification report of the model is:", crgrid)

===============================Accuracy score=======================================================
Accuracy score of the model is: 98.64603481624758
===============================Confusion matrix=======================================================
Confusion matrix of the model is: [[901   2]
 [ 12 119]]
===============================Classification report=======================================================
Classification report of the model is:               precision    recall  f1-score   support

           0       0.99      1.00      0.99       903
           1       0.98      0.91      0.94       131

    accuracy                           0.99      1034
   macro avg       0.99      0.95      0.97      1034
weighted avg       0.99      0.99      0.99      1034



#### OBSERVATIONS:

1.  The accuracy of email spam classification using Multinomial Naive Bayes Algorithm has increased from 98.45 to 98.64 using Hyper parameter tuning using Grid Search CV.

#### Step 13: Predict the type of email text based on the sample data

In [50]:
#### define the sample messages

sample_messages = [
    "Congratulations! You've won a free iPhone. Click here to claim.",
    "Hey, are we still meeting for lunch today?"
]

print(sample_messages)


#### perform the prediction on the vectorized data

prediction = best_estimator.predict(sample_messages)

print(prediction)


print("==========================First Sentence===============================================")

if(prediction[0] == 1):
    print("The email message is Spam")
else:
    print("The email message is Ham")


print("==========================Second Sentence===============================================")

if(prediction[1] == 1):
    print("The email message is Spam")
else:
    print("The email message is Ham")

["Congratulations! You've won a free iPhone. Click here to claim.", 'Hey, are we still meeting for lunch today?']
[1 0]
==========================First Sentence===============================================
The email message is Spam
==========================Second Sentence===============================================
The email message is Ham
